plugin/restore/mysql/mysql.go

```diff
@@ -561,49 +697,45 @@ func (r *Restore) downloadBinlogFile(ctx context.Context, binlog mysql.BinlogFil
+	log.Debug("Checking downloaded binlog file stat", zap.String("path", resultFilePath))
 	fileInfo, err := os.Stat(resultFilePath)
 	if err != nil {
-		_ = os.Remove(resultFilePath)
-		return fmt.Errorf("cannot stat %s, error: %w", resultFilePath, err)
+		return fmt.Errorf("cannot get file[%s] stat, error[%w]", resultFilePath, err)
 	}
 	if !isLast {
 		// Case 1: It's an archived binlog file, and we must ensure the file size equals what we queried from the MySQL server earlier.
-		if fileInfo.Size() != binlog.Size {
-			log.Error("Downloaded binlog file size does not match size queried on the MySQL server",
-				zap.String("binlog", binlog.Name),
-				zap.Int64("sizeInfo", binlog.Size),
+		if fileInfo.Size() != binlogFileToDownload.Size {
+			log.Error("Downloaded binlog file size does not equal to size queried on the MySQL server",
+				zap.String("binlog", binlogFileToDownload.Name),
+				zap.Int64("sizeInfo", binlogFileToDownload.Size),
 				zap.Int64("downloadedSize", fileInfo.Size()),
 			)
-			_ = os.Remove(resultFilePath)
-			return fmt.Errorf("downloaded binlog file[%s] size[%d] does not equal to size[%d] queried on MySQL server earlier", resultFilePath, fileInfo.Size(), binlog.Size)
+			return fmt.Errorf("downloaded binlog file[%s] size[%d] does not equal to size[%d] queried on MySQL server earlier", resultFilePath, fileInfo.Size(), binlogFileToDownload.Size)
 		}
 	} else {
 		// Case 2: It's the last binlog file we need (contains the targetTs).
 		// If it's the last (incomplete) binlog file on the MySQL server, it will grow as new writes hit the database server.
 		// We just need to check that the downloaded file size >= queried size, so it contains the targetTs event.
-		if fileInfo.Size() < binlog.Size {
+		if fileInfo.Size() < binlogFileToDownload.Size {
 			log.Error("Downloaded latest binlog file size is smaller than size queried on the MySQL server",
-				zap.String("binlog", binlog.Name),
-				zap.Int64("sizeInfo", binlog.Size),
+				zap.String("binlog", binlogFileToDownload.Name),
+				zap.Int64("sizeInfo", binlogFileToDownload.Size),
 				zap.Int64("downloadedSize", fileInfo.Size()),
 			)
-			_ = os.Remove(resultFilePath)
-			return fmt.Errorf("downloaded latest binlog file[%s] size[%d] is smaller than size[%d] queried on MySQL server earlier", resultFilePath, fileInfo.Size(), binlog.Size)
+			return fmt.Errorf("downloaded latest binlog file[%s] size[%d] is smaller than size[%d] queried on MySQL server earlier", resultFilePath, fileInfo.Size(), binlogFileToDownload.Size)
 		}
 	}
 
 	return nil
 }
 
-// getBinlogFilesMetaOnServer returns the metadata of binlog files.
-func (r *Restore) getBinlogFilesMetaOnServer(ctx context.Context) ([]mysql.BinlogFile, error) {
+// GetSortedBinlogFilesMetaOnServer returns the metadata of binlog files in ascending order by their numeric extension.
+func (r *Restore) GetSortedBinlogFilesMetaOnServer(ctx context.Context) ([]BinlogFile, error) {
```

Comment:
The caller assumes the returned list is continuous, if that's the case, this method should check that.

Reply:
It does not need to be continuous.
Imagine that there's a gap in the server binlog file list, but combined with the local binlog files, it is continuous. Then we can replay it to targetTs successfully too.

Reply:
I see